File: test/cs/Ssh.Test/PipeTests.cs
using System;
using System.IO;
using System.Text;
using System.Threading;
using System.Threading.Tasks;
using Microsoft.DevTunnels.Ssh.Events;
using Microsoft.DevTunnels.Ssh.IO;
using Microsoft.DevTunnels.Ssh.Messages;
using Xunit;

namespace Microsoft.DevTunnels.Ssh.Test
{
	public class PipeTests
	{
		private static readonly TimeSpan Timeout = TimeSpan.FromSeconds(10);

		private readonly SessionPair sessionPair1;
		private readonly SshClientSession clientSession1;
		private readonly SshServerSession serverSession1;
		private readonly SessionPair sessionPair2;
		private readonly SshClientSession clientSession2;
		private readonly SshServerSession serverSession2;

		public PipeTests()
		{
			var config = new SshSessionConfiguration();

			this.sessionPair1 = new SessionPair(config);
			this.clientSession1 = sessionPair1.ClientSession;
			this.serverSession1 = sessionPair1.ServerSession;
			this.sessionPair2 = new SessionPair(config);
			this.clientSession2 = sessionPair2.ClientSession;
			this.serverSession2 = sessionPair2.ServerSession;
		}

		private async Task ConnectSessions()
		{
			await Task.WhenAll(
				this.sessionPair1.ConnectAsync(true),
				this.sessionPair2.ConnectAsync(true));
		}

		[Theory]
		[InlineData(false)]
		[InlineData(true)]
		public async Task PipeChannelClose(bool closeTarget)
		{
			await this.ConnectSessions();
			var (clientChannel1, serverChannel1) = await this.sessionPair1.OpenChannelAsync();
			var (clientChannel2, serverChannel2) = await this.sessionPair2.OpenChannelAsync();
			var pipeTask = serverChannel1.PipeAsync(serverChannel2);

			var closedCompletion = new TaskCompletionSource<SshChannelClosedEventArgs>();
			(closeTarget ? clientChannel1 : clientChannel2).Closed +=
				(sender, e) => closedCompletion.TrySetResult(e);
			await (closeTarget ? clientChannel2 : clientChannel1).CloseAsync();
			await closedCompletion.Task.WithTimeout(Timeout);
			await pipeTask.WithTimeout(Timeout);
		}

		[Theory]
		[InlineData(false)]
		[InlineData(true)]
		public async Task PipeChannelSend(bool fromTarget)
		{
			await this.ConnectSessions();
			var (clientChannel1, serverChannel1) = await this.sessionPair1.OpenChannelAsync();
			var (clientChannel2, serverChannel2) = await this.sessionPair2.OpenChannelAsync();
			_ = serverChannel1.PipeAsync(serverChannel2);

			var testData = Buffer.From(Encoding.UTF8.GetBytes("test"));
			var dataCompletion = new TaskCompletionSource<Buffer>();
			(fromTarget ? clientChannel1 : clientChannel2).DataReceived +=
				(sender, data) => dataCompletion.TrySetResult(data.Copy());
			await (fromTarget ? clientChannel2 : clientChannel1).SendAsync(
				testData, CancellationToken.None);
			var receivedData = await dataCompletion.Task.WithTimeout(Timeout);
			Assert.True(receivedData.Equals(testData));
		}

		[Theory]
		[InlineData(false)]
		[InlineData(true)]
		public async Task PipeChannelSendSequence(bool fromTarget)
		{
			await this.ConnectSessions();
			var (clientChannel1, serverChannel1) = await this.sessionPair1.OpenChannelAsync();
			var (clientChannel2, serverChannel2) = await this.sessionPair2.OpenChannelAsync();
			_ = serverChannel1.PipeAsync(serverChannel2);

			int count = 1000;
			var receivedCompletion = new TaskCompletionSource<int>();
			int receivedCount = 0;
			(fromTarget ? clientChannel1 : clientChannel2).DataReceived += (sender, data) =>
			{
				var expectedData = Buffer.From(Encoding.UTF8.GetBytes("test:" + receivedCount));
				Assert.True(data.Equals(expectedData));
				if (++receivedCount == count)
				{
					receivedCompletion.SetResult(count);
				}
				((SshChannel)sender).AdjustWindow((uint)data.Count);
			};

			for (int i = 0; i < count; i++)
			{
				var testData = Buffer.From(Encoding.UTF8.GetBytes("test:" + i));
				_ = (fromTarget ? clientChannel2 : clientChannel1).SendAsync(
					testData, CancellationToken.None);
			}

			await receivedCompletion.Task.WithTimeout(Timeout);
		}

		[Theory]
		[InlineData(false)]
		[InlineData(true)]
		public async Task PipeChannelSendLargeData(bool fromTarget)
		{
			await this.ConnectSessions();
			var (clientChannel1, serverChannel1) = await this.sessionPair1.OpenChannelAsync();
			var (clientChannel2, serverChannel2) = await this.sessionPair2.OpenChannelAsync();
			_ = serverChannel1.PipeAsync(serverChannel2);

			// Test data that is larger than the channel flow-control window size (1MB).
			const int largeDataSize = 1024 * 1024 * 7 / 2;
			byte[] largeData = new byte[largeDataSize];
			for (int i = 0; i < largeData.Length; i++) largeData[i] = (byte)(i & 0xFF);

			var receivingStream = new MemoryStream();
			var receivedCompletion = new TaskCompletionSource<byte[]>();

			(fromTarget ? clientChannel1 : clientChannel2).DataReceived += (sender, data) =>
			{
				receivingStream.Write(data.Array, data.Offset, data.Count);
				if (receivingStream.Length >= largeDataSize)
				{
					receivedCompletion.SetResult(receivingStream.ToArray());
				}

				((SshChannel)sender).AdjustWindow((uint)data.Count);
			};

			await (fromTarget ? clientChannel2 : clientChannel1).SendAsync(
				largeData, CancellationToken.None).WithTimeout(Timeout);
			var receivedData = await receivedCompletion.Task.WithTimeout(Timeout);
			Assert.True(Buffer.From(receivedData).Equals(largeData));
		}

		[Fact]
		public async Task PipeChannelPendingRequest()
		{
			await this.ConnectSessions();
			var (clientChannel1, serverChannel1) = await this.sessionPair1.OpenChannelAsync();
			var (clientChannel2, serverChannel2) = await this.sessionPair2.OpenChannelAsync();
			_ = serverChannel1.PipeAsync(clientChannel2);

			var firstRequest = new ChannelRequestMessage { RequestType = "first", WantReply = true };
			var secondRequest = new ChannelRequestMessage { RequestType = "second", WantReply = true };
			TaskCompletionSource<bool> secondMessageCompletion = new TaskCompletionSource<bool>();
			TaskCompletionSource<bool> firstMessageCompletion = new TaskCompletionSource<bool>();
			var responseTask = async (SshRequestEventArgs<ChannelRequestMessage> e) =>
			{
				if (e.Request.RequestType == "second")
				{
					secondMessageCompletion.SetResult(true);
				}
				else
				{
					firstMessageCompletion.SetResult(true);
					await secondMessageCompletion.Task;
				}
				return new ChannelSuccessMessage() as SshMessage;
			};
			serverChannel2.Request += (sender, e) =>
			{
				e.ResponseTask = responseTask(e);
			};
			var firstTask = clientChannel1.RequestAsync(
				firstRequest, CancellationToken.None);
			await clientChannel1.RequestAsync(
				secondRequest, CancellationToken.None);
			await firstTask;
			Assert.True(await secondMessageCompletion.Task);
			Assert.True(await firstMessageCompletion.Task);
		}

		[Theory]
		[InlineData(false)]
		[InlineData(true)]
		public async Task PipeSessionClose(bool closeTarget)
		{
			await this.ConnectSessions();
			var pipeTask = this.sessionPair1.ServerSession.PipeAsync(this.sessionPair2.ServerSession);

			var closedCompletion = new TaskCompletionSource<SshSessionClosedEventArgs>();
			(closeTarget ? this.sessionPair1.ClientSession : this.sessionPair2.ClientSession).Closed +=
				(sender, e) => closedCompletion.TrySetResult(e);
			await (closeTarget ? this.sessionPair2.ClientSession : this.sessionPair1.ClientSession)
				.CloseAsync(SshDisconnectReason.ByApplication);

			var closedEvent = await closedCompletion.Task.WithTimeout(Timeout);
			Assert.Equal(SshDisconnectReason.ByApplication, closedEvent.Reason);
			await pipeTask.WithTimeout(Timeout);
		}

		[Theory]
		[InlineData(false)]
		[InlineData(true)]
		public async Task PipeSessionChannelOpen(bool fromTarget)
		{
			await this.ConnectSessions();
			_ = this.sessionPair1.ServerSession.PipeAsync(this.sessionPair2.ServerSession);

			var channelTask =
				(fromTarget ? this.sessionPair1.ClientSession : this.sessionPair2.ClientSession)
				.AcceptChannelAsync("test");
			_ = await (fromTarget ? this.sessionPair2.ClientSession : this.sessionPair1.ClientSession)
				.OpenChannelAsync("test").WithTimeout(Timeout);
			var channel = await channelTask.WithTimeout(Timeout);
			Assert.Equal("test", channel.ChannelType);
		}

		[Theory]
		[InlineData(false)]
		[InlineData(true)]
		public async Task PipeSessionChannelOpenAndClose(bool fromTarget)
		{
			await this.ConnectSessions();
			_ = this.sessionPair1.ServerSession.PipeAsync(this.sessionPair2.ServerSession);

			var channelTask =
				(fromTarget ? this.sessionPair1.ClientSession : this.sessionPair2.ClientSession)
				.AcceptChannelAsync();
			var channelA = await
				(fromTarget ? this.sessionPair2.ClientSession : this.sessionPair1.ClientSession)
				.OpenChannelAsync().WithTimeout(Timeout);
			var channelB = await channelTask.WithTimeout(Timeout);

			var closedCompletion = new TaskCompletionSource<SshChannelClosedEventArgs>();
			(fromTarget ? channelA : channelB).Closed +=
				(sender, e) => closedCompletion.TrySetResult(e);
			await (fromTarget ? channelB : channelA).CloseAsync();
			await closedCompletion.Task.WithTimeout(Timeout);
		}

		[Theory]
		[InlineData(false)]
		[InlineData(true)]
		public async Task PipeSessionChannelSend(bool fromTarget)
		{
			await this.ConnectSessions();
			_ = this.sessionPair1.ServerSession.PipeAsync(this.sessionPair2.ServerSession);

			var channelTask =
				(fromTarget ? this.sessionPair1.ClientSession : this.sessionPair2.ClientSession)
				.AcceptChannelAsync();
			var channelA = await
				(fromTarget ? this.sessionPair2.ClientSession : this.sessionPair1.ClientSession)
				.OpenChannelAsync().WithTimeout(Timeout);
			var channelB = await channelTask.WithTimeout(Timeout);

			var testData = Buffer.From(Encoding.UTF8.GetBytes("test"));
			var dataCompletion = new TaskCompletionSource<Buffer>();
			channelB.DataReceived += (sender, data) =>
			{
				dataCompletion.TrySetResult(data.Copy());
			};
			await channelA.SendAsync(testData, CancellationToken.None);
			var receivedData = await dataCompletion.Task.WithTimeout(Timeout);
			Assert.True(receivedData.Equals(testData));
		}

		[Fact]
		public async Task PipeExtensibleSessionRequest()
		{
			await this.ConnectSessions();
			_ = this.sessionPair1.ServerSession.PipeAsync(this.sessionPair2.ServerSession);

			var client1 = this.sessionPair1.ClientSession;
			var client2 = this.sessionPair2.ClientSession;

			var requestCompletion = new TaskCompletionSource<SessionRequestMessage>();
			client2.Request += (sender, e) =>
			{
				requestCompletion.TrySetResult(e.Request);
				e.ResponseTask = Task.FromResult<SshMessage>(new TestSessionRequestSuccessMessage());
			};

			var requestTask = client1.RequestAsync<TestSessionRequestSuccessMessage>(
				new TestSessionRequestMessage { RequestType = "test" }, CancellationToken.None);

			var testRequest = await requestCompletion.Task.WithTimeout(Timeout);
			Assert.Equal("test", testRequest.RequestType);
			testRequest.ConvertTo<TestSessionRequestMessage>();

			var testResponse = await requestTask;
			Assert.NotNull(testResponse);
		}

		[Fact]
		public async Task PipeSessionPendingRequest()
		{
			await this.ConnectSessions();
			_ = this.sessionPair1.ServerSession.PipeAsync(this.sessionPair2.ClientSession);

			var firstRequest = new SessionRequestMessage { RequestType = "first", WantReply = true };
			var secondRequest = new SessionRequestMessage { RequestType = "second", WantReply = true };
			TaskCompletionSource<bool> secondMessageCompletion = new TaskCompletionSource<bool>();
			TaskCompletionSource<bool> firstMessageCompletion = new TaskCompletionSource<bool>();
			var responseTask = async (SshRequestEventArgs<SessionRequestMessage> e) =>
			{
				if (e.Request.RequestType == "second")
				{
					secondMessageCompletion.SetResult(true);
				}
				else
				{
					firstMessageCompletion.SetResult(true);
					await secondMessageCompletion.Task;
				}
				return new SessionRequestSuccessMessage() as SshMessage;
			};
			this.sessionPair2.ServerSession.Request += (sender, e) =>
			{
				e.ResponseTask = responseTask(e);
			};
			var firstTask = this.sessionPair1.ClientSession.RequestAsync(
				firstRequest, CancellationToken.None);
			await this.sessionPair1.ClientSession.RequestAsync(
				secondRequest, CancellationToken.None);
			await firstTask;
			Assert.True(await secondMessageCompletion.Task);
			Assert.True(await firstMessageCompletion.Task);
		}

		[Fact]
		public async Task PipeExtensibleChannelOpen()
		{
			await this.ConnectSessions();
			_ = this.sessionPair1.ServerSession.PipeAsync(this.sessionPair2.ServerSession);

			var requestCompletion = new TaskCompletionSource<ChannelOpenMessage>();
			var channelCompletion = new TaskCompletionSource<SshChannel>();
			this.sessionPair2.ClientSession.ChannelOpening += (sender, e) =>
			{
				requestCompletion.TrySetResult(e.Request);
				channelCompletion.TrySetResult(e.Channel);
			};

			var openTask = this.sessionPair1.ClientSession.OpenChannelAsync(
				new TestChannelOpenMessage { ChannelType = "test" }, null).WithTimeout(Timeout);

			var testRequest = await requestCompletion.Task.WithTimeout(Timeout);
			testRequest.ConvertTo<TestChannelOpenMessage>();

			var channel1 = await openTask.WithTimeout(Timeout);
			Assert.Equal("test", channel1.ChannelType);

			var channel2 = await channelCompletion.Task;
			Assert.Equal("test", channel2.ChannelType);
		}

		[Theory]
		[InlineData(false)]
		[InlineData(true)]
		public async Task PipeExtensibleChannelRequest(bool withChannelIdMapping)
		{
			await this.ConnectSessions();

			if (withChannelIdMapping)
			{
				// Open a channel BEFORE piping, so that the channel IDs will not be in sync.
				// Channel piping should support re-mapping channel IDs.
				_ = await this.sessionPair1.ClientSession.OpenChannelAsync();
			}

			_ = this.sessionPair1.ServerSession.PipeAsync(this.sessionPair2.ServerSession);

			var acceptTask = this.sessionPair2.ClientSession.AcceptChannelAsync("test");
			var channel1 = await this.sessionPair1.ClientSession.OpenChannelAsync("test")
				.WithTimeout(Timeout);
			var channel2 = await acceptTask.WithTimeout(Timeout);

			var requestCompletion = new TaskCompletionSource<ChannelRequestMessage>();
			channel2.Request += (sender, e) =>
			{
				requestCompletion.TrySetResult(e.Request);
				e.IsAuthorized = true;
			};

			var requestTask = channel1.RequestAsync(
				new TestChannelRequestMessage { RequestType = "test" });

			var request = await requestCompletion.Task.WithTimeout(Timeout);
			Assert.Equal("test", request.RequestType);
			request.ConvertTo<TestChannelRequestMessage>();

			var result = await requestTask.WithTimeout(Timeout);
			Assert.True(result);
		}

		private class TestSessionRequestMessage : SessionRequestMessage
		{
			protected override void OnWrite(ref SshDataWriter writer)
			{
				base.OnWrite(ref writer);
				writer.Write((uint)1);
			}

			protected override void OnRead(ref SshDataReader reader)
			{
				base.OnRead(ref reader);
				Assert.Equal((uint)1, reader.ReadUInt32());
			}
		}

		private class TestSessionRequestSuccessMessage : SessionRequestSuccessMessage
		{
			protected override void OnWrite(ref SshDataWriter writer)
			{
				base.OnWrite(ref writer);
				writer.Write((uint)1);
			}

			protected override void OnRead(ref SshDataReader reader)
			{
				base.OnRead(ref reader);
				Assert.Equal((uint)1, reader.ReadUInt32());
			}
		}

		private class TestChannelOpenMessage : ChannelOpenMessage
		{
			protected override void OnWrite(ref SshDataWriter writer)
			{
				base.OnWrite(ref writer);
				writer.Write((uint)1);
			}

			protected override void OnRead(ref SshDataReader reader)
			{
				base.OnRead(ref reader);
				Assert.Equal((uint)1, reader.ReadUInt32());
			}
		}

		private class TestChannelRequestMessage : ChannelRequestMessage
		{
			protected override void OnWrite(ref SshDataWriter writer)
			{
				base.OnWrite(ref writer);
				writer.Write((uint)1);
			}

			protected override void OnRead(ref SshDataReader reader)
			{
				base.OnRead(ref reader);
				Assert.Equal((uint)1, reader.ReadUInt32());
			}
		}
	}
}

File: test/ts/ssh-test/pipeTests.ts
//
//  Copyright (c) Microsoft Corporation. All rights reserved.
//

import * as assert from 'assert';
import * as net from 'net';
import { suite, test, params, pending, slow, timeout } from '@testdeck/mocha';

import {
	ChannelOpenMessage,
	ChannelRequestMessage,
	PromiseCompletionSource,
	SessionRequestFailureMessage,
	SessionRequestMessage,
	SessionRequestSuccessMessage,
	SshAlgorithms,
	SshChannel,
	SshChannelClosedEventArgs,
	SshChannelError,
	SshChannelOpenFailureReason,
	SshClientSession,
	SshDataReader,
	SshDataWriter,
	SshDisconnectReason,
	SshServerSession,
	SshSessionClosedEventArgs,
} from '@microsoft/dev-tunnels-ssh';
import { connectSessionPair, createSessionPair, openChannel } from './sessionPair';
import { expectError, until, withTimeout } from './promiseUtils';

const timeoutMs = 5000;

class TestSessionRequestMessage extends SessionRequestMessage {
	protected onWrite(writer: SshDataWriter): void {
		super.onWrite(writer);
		writer.writeUInt32(1);
	}

	protected onRead(reader: SshDataReader) {
		super.onRead(reader);
		assert.equal(1, reader.readUInt32());
	}
}

class TestSessionRequestSuccessMessage extends SessionRequestSuccessMessage {
	protected onWrite(writer: SshDataWriter): void {
		super.onWrite(writer);
		writer.writeUInt32(1);
	}

	protected onRead(reader: SshDataReader) {
		super.onRead(reader);
		assert.equal(1, reader.readUInt32());
	}
}

class TestChannelOpenMessage extends ChannelOpenMessage {
	protected onWrite(writer: SshDataWriter): void {
		super.onWrite(writer);
		writer.writeUInt32(1);
	}

	protected onRead(reader: SshDataReader) {
		super.onRead(reader);
		assert.equal(1, reader.readUInt32());
	}
}

class TestChannelRequestMessage extends ChannelRequestMessage {
	protected onWrite(writer: SshDataWriter): void {
		super.onWrite(writer);
		writer.writeUInt32(1);
	}

	protected onRead(reader: SshDataReader) {
		super.onRead(reader);
		assert.equal(1, reader.readUInt32());
	}
}

@suite
@slow(4000)
@timeout(2 * timeoutMs)
export class PipeTests {
	private clientSession1!: SshClientSession;
	private serverSession1!: SshServerSession;
	private clientSession2!: SshClientSession;
	private serverSession2!: SshServerSession;

	public after(): void {
		this.clientSession1?.dispose();
		this.serverSession1?.dispose();
		this.clientSession2?.dispose();
		this.serverSession2?.dispose();
	}

	private async createSessions(): Promise<void> {
		[this.clientSession1, this.serverSession1] = await createSessionPair();
		[this.clientSession2, this.serverSession2] = await createSessionPair();

		const serverKey = await SshAlgorithms.publicKey.ecdsaSha2Nistp384!.generateKeyPair();
		this.serverSession1.credentials.publicKeys = [serverKey];
		this.serverSession2.credentials.publicKeys = [serverKey];
	}

	@test
	@params({ closeTarget: true })
	@params({ closeTarget: false })
	@params.naming((p) => `pipeChannelClose(closeTarget=${p.closeTarget})`)
	public async pipeChannelClose({ closeTarget }: { closeTarget: boolean }): Promise<void> {
		await this.createSessions();
		await connectSessionPair(this.clientSession1, this.serverSession1);
		await connectSessionPair(this.clientSession2, this.serverSession2);
		const [clientChannel1, serverChannel1] = await openChannel(
			this.clientSession1,
			this.serverSession1,
		);
		const [clientChannel2, serverChannel2] = await openChannel(
			this.clientSession2,
			this.serverSession2,
		);
		const pipePromise = serverChannel1.pipe(serverChannel2);

		const closedCompletion = new PromiseCompletionSource<SshChannelClosedEventArgs>();
		(closeTarget ? clientChannel1 : clientChannel2).onClosed((e) => {
			closedCompletion.resolve(e);
		});
		await (closeTarget ? clientChannel2 : clientChannel1).close();
		await withTimeout(closedCompletion.promise, timeoutMs);
		await withTimeout(pipePromise, timeoutMs);
	}

	@test
	@params({ fromTarget: true })
	@params({ fromTarget: false })
	@params.naming((p) => `pipeChannelSend(fromTarget=${p.fromTarget})`)
	public async pipeChannelSend({ fromTarget }: { fromTarget: boolean }): Promise<void> {
		await this.createSessions();
		await connectSessionPair(this.clientSession1, this.serverSession1);
		await connectSessionPair(this.clientSession2, this.serverSession2);
		const [clientChannel1, serverChannel1] = await openChannel(
			this.clientSession1,
			this.serverSession1,
		);
		const [clientChannel2, serverChannel2] = await openChannel(
			this.clientSession2,
			this.serverSession2,
		);
		const pipePromise = serverChannel1.pipe(serverChannel2);

		const testData = Buffer.from('test', 'utf8');
		const dataCompletion = new PromiseCompletionSource<Buffer>();
		(fromTarget ? clientChannel1 : clientChannel2).onDataReceived((data) => {
			dataCompletion.resolve(data);
		});
		await (fromTarget ? clientChannel2 : clientChannel1).send(testData);
		const receivedData = await withTimeout(dataCompletion.promise, timeoutMs);
		assert(receivedData.equals(testData));
	}

	@test
	@params({ fromTarget: true })
	@params({ fromTarget: false })
	@params.naming((p) => `pipeChannelSendSequence(fromTarget=${p.fromTarget})`)
	public async pipeChannelSendSequence({ fromTarget }: { fromTarget: boolean }): Promise<void> {
		await this.createSessions();
		await connectSessionPair(this.clientSession1, this.serverSession1);
		await connectSessionPair(this.clientSession2, this.serverSession2);
		const [clientChannel1, serverChannel1] = await openChannel(
			this.clientSession1,
			this.serverSession1,
		);
		const [clientChannel2, serverChannel2] = await openChannel(
			this.clientSession2,
			this.serverSession2,
		);
		const pipePromise = serverChannel1.pipe(serverChannel2);

		const count = 1000;
		const receivedCompletion = new PromiseCompletionSource<void>();
		let receivedCount = 0;
		(fromTarget ? clientChannel1 : clientChannel2).onDataReceived((data) => {
			const expectedData = Buffer.from('test:' + receivedCount, 'utf8');
			assert(data.equals(expectedData));
			if (++receivedCount === count) {
				receivedCompletion.resolve();
			}
			(fromTarget ? clientChannel1 : clientChannel2).adjustWindow(data.length);
		});

		for (let i = 0; i < count; i++) {
			const testData = Buffer.from('test:' + i, 'utf8');
			const _ = (fromTarget ? clientChannel2 : clientChannel1).send(testData);
		}

		await withTimeout(receivedCompletion.promise, timeoutMs);
	}

	@test
	@params({ fromTarget: true })
	@params({ fromTarget: false })
	@params.naming((p) => `pipeChannelSendLargeData(fromTarget=${p.fromTarget})`)
	public async pipeChannelSendLargeData({ fromTarget }: { fromTarget: boolean }): Promise<void> {
		await this.createSessions();
		await connectSessionPair(this.clientSession1, this.serverSession1);
		await connectSessionPair(this.clientSession2, this.serverSession2);
		const [clientChannel1, serverChannel1] = await openChannel(
			this.clientSession1,
			this.serverSession1,
		);
		const [clientChannel2, serverChannel2] = await openChannel(
			this.clientSession2,
			this.serverSession2,
		);
		const pipePromise = serverChannel1.pipe(serverChannel2);

		// Test data that is larger than the channel flow-control window size (1MB).
		const largeDataSize = (1024 * 1024 * 7) / 2;
		const largeData = Buffer.alloc(largeDataSize);
		for (let i = 0; i < largeData.length; i++) largeData[i] = i & 0xff;

		let receivingBuffer = Buffer.alloc(0);
		const receivedCompletion = new PromiseCompletionSource<Buffer>();

		(fromTarget ? clientChannel1 : clientChannel2).onDataReceived((data) => {
			receivingBuffer = Buffer.concat([receivingBuffer, data]);
			if (receivingBuffer.length >= largeDataSize) {
				receivedCompletion.resolve(receivingBuffer);
			}

			(fromTarget ? clientChannel1 : clientChannel2).adjustWindow(data.length);
		});

		await (fromTarget ? clientChannel2 : clientChannel1).send(largeData);
		const receivedData = await withTimeout(receivedCompletion.promise, timeoutMs);
		assert(receivedData.equals(largeData));
	}

	@test
	@params({ closeTarget: true })
	@params({ closeTarget: false })
	@params.naming((p) => `pipeSessionClose(closeTarget=${p.closeTarget})`)
	public async pipeSessionClose({ closeTarget }: { closeTarget: boolean }): Promise<void> {
		await this.createSessions();
		await connectSessionPair(this.clientSession1, this.serverSession1);
		await connectSessionPair(this.clientSession2, this.serverSession2);
		const pipePromise = this.serverSession1.pipe(this.serverSession2);

		const closedCompletion = new PromiseCompletionSource<SshSessionClosedEventArgs>();
		(closeTarget ? this.clientSession1 : this.clientSession2).onClosed((e) => {
			closedCompletion.resolve(e);
		});
		await (closeTarget ? this.clientSession2 : this.clientSession1).close(
			SshDisconnectReason.byApplication,
		);

		const closedEvent = await withTimeout(closedCompletion.promise, timeoutMs);
		assert.equal(SshDisconnectReason.byApplication, closedEvent.reason);
		await withTimeout(pipePromise, timeoutMs);
	}

	@test
	@params({ fromTarget: true })
	@params({ fromTarget: false })
	@params.naming((p) => `pipeSessionChannelOpen(fromTarget=${p.fromTarget})`)
	public async pipeSessionChannelOpen({ fromTarget }: { fromTarget: boolean }): Promise<void> {
		await this.createSessions();
		await connectSessionPair(this.clientSession1, this.serverSession1);
		await connectSessionPair(this.clientSession2, this.serverSession2);
		const pipePromise = this.serverSession1.pipe(this.serverSession2);

		const channelPromise = (fromTarget ? this.clientSession1 : this.clientSession2).acceptChannel(
			'test',
		);
		await (fromTarget ? this.clientSession2 : this.clientSession1).openChannel('test');
		const channel = await withTimeout(channelPromise, timeoutMs);
		assert.equal('test', channel.channelType);
	}

	@test
	@params({ fromTarget: true })
	@params({ fromTarget: false })
	@params.naming((p) => `pipeSessionChannelOpenAndClose(fromTarget=${p.fromTarget})`)
	public async pipeSessionChannelOpenAndClose({
		fromTarget,
	}: {
		fromTarget: boolean;
	}): Promise<void> {
		await this.createSessions();
		await connectSessionPair(this.clientSession1, this.serverSession1);
		await connectSessionPair(this.clientSession2, this.serverSession2);
		const pipePromise = this.serverSession1.pipe(this.serverSession2);

		const channelPromise = (fromTarget ? this.clientSession1 : this.clientSession2).acceptChannel(
			'test',
		);
		const channelA = await (fromTarget ? this.clientSession2 : this.clientSession1).openChannel(
			'test',
		);
		const channelB = await withTimeout(channelPromise, timeoutMs);

		const closedCompletion = new PromiseCompletionSource<SshChannelClosedEventArgs>();
		(fromTarget ? channelA : channelB).onClosed((e) => {
			closedCompletion.resolve(e);
		});
		await (fromTarget ? channelB : channelA).close();
		await withTimeout(closedCompletion.promise, timeoutMs);
	}

	@test
	@params({ fromTarget: true })
	@params({ fromTarget: false })
	@params.naming((p) => `pipeSessionChannelSend(fromTarget=${p.fromTarget})`)
	public async pipeSessionChannelSend({ fromTarget }: { fromTarget: boolean }): Promise<void> {
		await this.createSessions();
		await connectSessionPair(this.clientSession1, this.serverSession1);
		await connectSessionPair(this.clientSession2, this.serverSession2);
		const pipePromise = this.serverSession1.pipe(this.serverSession2);

		const channelPromise = (fromTarget ? this.clientSession1 : this.clientSession2).acceptChannel(
			'test',
		);
		const channelA = await (fromTarget ? this.clientSession2 : this.clientSession1).openChannel(
			'test',
		);
		const channelB = await withTimeout(channelPromise, timeoutMs);

		const testData = Buffer.from('test', 'utf8');
		const dataCompletion = new PromiseCompletionSource<Buffer>();
		(fromTarget ? channelA : channelB).onDataReceived((data) => {
			dataCompletion.resolve(data);
		});
		await (fromTarget ? channelB : channelA).send(testData);
		const receivedData = await withTimeout(dataCompletion.promise, timeoutMs);
		assert(receivedData.equals(testData));
	}

	@test
	public async pipeExtensibleSessionRequest() {
		await this.createSessions();
		await connectSessionPair(this.clientSession1, this.serverSession1);
		await connectSessionPair(this.clientSession2, this.serverSession2);
		const pipePromise = this.serverSession1.pipe(this.serverSession2);

		const requestCompletion = new PromiseCompletionSource<SessionRequestMessage>();
		this.clientSession2.onRequest((e) => {
			requestCompletion.resolve(e.request);
			e.responsePromise = Promise.resolve(new TestSessionRequestSuccessMessage());
		});

		const request = new TestSessionRequestMessage();
		request.requestType = 'test';
		const requestTask = this.clientSession1.requestResponse(
			request,
			TestSessionRequestSuccessMessage,
			SessionRequestFailureMessage,
		);

		const testRequest = await withTimeout(requestCompletion.promise, timeoutMs);
		assert.equal('test', testRequest.requestType);
		testRequest.convertTo(new TestSessionRequestMessage());

		const testResponse = await requestTask;
		assert(testResponse);
	}

	@test
	public async pipeExtensibleSessionMultipleRequestNoReply() {
		await this.createSessions();
		await connectSessionPair(this.clientSession1, this.serverSession1);
		await connectSessionPair(this.clientSession2, this.serverSession2);
		const pipePromise = this.serverSession1.pipe(this.clientSession2);

		for (let i = 0; i < 3; i++) {
			const requestCompletion = new PromiseCompletionSource<SessionRequestMessage>();
			const toDispose = this.clientSession1.onRequest((e) => {
				requestCompletion.resolve(e.request);
				e.isAuthorized = true;
			});

			const request = new TestSessionRequestMessage();
			request.requestType = 'test';
			const requestTask = this.serverSession2.request(
				request
			);

			const testRequest = await withTimeout(requestCompletion.promise, timeoutMs);
			assert.equal('test', testRequest.requestType);
			testRequest.convertTo(new TestSessionRequestMessage());

			const testResponse = await requestTask;
			assert(testResponse);

			toDispose.dispose();
		}
	}

	@test
	public async pipeExtensibleChannelOpen() {
		await this.createSessions();
		await connectSessionPair(this.clientSession1, this.serverSession1);
		await connectSessionPair(this.clientSession2, this.serverSession2);
		const pipePromise = this.serverSession1.pipe(this.serverSession2);

		const requestCompletion = new PromiseCompletionSource<ChannelOpenMessage>();
		const channelCompletion = new PromiseCompletionSource<SshChannel>();
		this.clientSession2.onChannelOpening((e) => {
			requestCompletion.resolve(e.request);
			channelCompletion.resolve(e.channel);
		});

		const openMessage = new TestChannelOpenMessage();
		openMessage.channelType = 'test';
		const openPromise = this.clientSession1.openChannel(openMessage);

		const testRequest = await withTimeout(requestCompletion.promise, timeoutMs);
		testRequest.convertTo(new TestChannelOpenMessage());

		const channel1 = await withTimeout(openPromise, timeoutMs);
		assert.equal(channel1.channelType, 'test');

		const channel2 = await withTimeout(channelCompletion.promise, timeoutMs);
		assert.equal(channel2.channelType, 'test');
	}

	@test
	@params({ withChannelIdMapping: false })
	@params({ withChannelIdMapping: true })
	@params.naming(
		(p) => `pipeExtensibleChannelRequest(withChannelIdMapping=${p.withChannelIdMapping})`,
	)
	public async pipeExtensibleChannelRequest({
		withChannelIdMapping,
	}: {
		withChannelIdMapping: boolean;
	}) {
		await this.createSessions();
		await connectSessionPair(this.clientSession1, this.serverSession1);
		await connectSessionPair(this.clientSession2, this.serverSession2);

		if (withChannelIdMapping) {
			// Open a channel BEFORE piping, so that the channel IDs will not be in sync.
			// Channel piping should support re-mapping channel IDs.
			const _ = await this.clientSession1.openChannel();
		}

		const pipePromise = this.serverSession1.pipe(this.serverSession2);

		const acceptPromise = this.clientSession2.acceptChannel('test');
		const channel1 = await withTimeout(this.clientSession1.openChannel('test'), timeoutMs);
		const channel2 = await withTimeout(acceptPromise, timeoutMs);

		const requestCompletion = new PromiseCompletionSource<ChannelRequestMessage>();
		channel2.onRequest((e) => {
			requestCompletion.resolve(e.request);
			e.isAuthorized = true;
		});

		const requestMessage = new TestChannelRequestMessage();
		requestMessage.requestType = 'test';
		const requestPromise = channel1.request(requestMessage);

		const request = await withTimeout(requestCompletion.promise, timeoutMs);
		assert.equal(request.requestType, 'test');
		request.convertTo(new TestChannelRequestMessage());

		const result = await withTimeout(requestPromise, timeoutMs);
		assert(result);
	}
}
